### DISCUSSION TITLE
(VB6) Send cookies to the webserver, with cHttpRequest ?

### DISCUSSION THREAD

**Discussion by CouinCouin** [wqweto/VbAsyncSocket#35] [General]
Hi Vladimir :) 

I use the cHttpRequest dll and I would know if it is possible to send cookies to the webserver from the VB6 Sub? 

Here is my VB6 code:
```
    Dim req As New cHttpRequest
	
    With req
        .SetTimeouts 500, 500, 500, 500
        .Open_ "POST", "http://192.168.0.250/forum/external.php", True
        .SetRequestHeader "Content-Type", "application/x-www-form-urlencoded; charset=""UTF-8"""
        .SetRequestHeader "Cookie", "cookietosend=" & cookievalue & "; Path=/"
        .Option_(WinHttpRequestOption_SslErrorIgnoreFlags) = 13056 '&H3300
        .Send "&username=" & username & _
            "&password=" & Password
        DoEvents
        .WaitForResponse
        Debug.Print .ResponseText
    End With
```

Here is my external.php code: 
```
   print("Cookies received by the server:\n\r");
   foreach ($_COOKIE as $k => $v) {
      print "   $k = $v\n\r";
   }
   print "End of php script";
```

The VB6 console show only:
```
Cookies received by the server:

End of php script
```

I tried _.SetRequestHeader "Set-Cookie"_ instead of _.SetRequestHeader "Cookie"_ but nothing better.

A small idea ? :)

Thanks ! :D
Couin

**Comment by wqweto** [comment]
`Host`, `Cookie` and `Content-Length` headers are controlled internally by `cHttpRequest` so any custom values set by `SetRequestHeader` are ignored (because of redirects handling).

You can tweak `cHttpRequest` source code by making function `pvSetCookie` public (like in latest commit 3ff1d44de9f9ee6f6ca45625dfb77a008387be8c) which will allow injecting custom cookies for each domain+path pair. This way instead of `.SetRequestHeader "Cookie", "cookietosend=" & cookievalue & "; Path=/"` one will be able to use `.SetCookie "192.168.0.250", "/", "cookietosend=" & cookievalue`

I'll include this tweak in next release too.

**Comment by CouinCouin** [comment]
Hi again :)

Thanks for your answer. 
Oh, so "Content-Type" SetRequestHeader is to remove also ? 

About tweak, why not, but I use the dll https://github.com/wqweto/VbAsyncSocket/releases/download/HRR-1.0.3/HttpRequest.zip you provided me few monthes ago. 
I did not updated to 1.0.4 and I see it is missing in 1.0.5.  I don't know how to create/modify a dll :(

**Reply by wqweto** [reply]
If you cannot tweak `SetCookie` in your setup you'll have to wait for the next binary release which will include the fix.

Btw, `Content-Type` is different than `Content-Length` so `Content-Type` header is fine/supported.

**Comment by CouinCouin** [comment]
Thanks :) 
Also, in the archive, if there are all files needed to build the dll, if you have some little guidelines ?

**Reply by wqweto** [reply]
The project to build `HttpRequest.dll` is in `contrib/dll` directory.

Keep in mind that you'll have to switch `cHttpRequest` class to public before building the binary.

**Comment by CouinCouin** [comment]
Thanks for the info :)
Can the DLL be buit from VB6 IDE?  Or should I need another software? 
Edit: I just saw that opening in the VB6 IDE, purpose in the menu, to Make the dll, but I get a "No creatable public component detected", even by changing to Public" instead of "Private this line :
`Private Property Get pvAddressOfReaderCallback() As cHttpRequest`

**Reply by wqweto** [reply]
You'll have to switch `cHttpRequest` class to public before building the binary.

**Reply by CouinCouin** [reply]
Oh sorry, I though about the procedure.
In the cclass properties, I have:
1- Private
2- PublicNotCreatable
5- MultiUse
6- GlobalMultiuse

3 and 4 choices are missing, I don' know if normal or not. Shoudl I should 5 or 6 instead ?
Thanks :)

**Reply by wqweto** [reply]
I just added precompiled [HttpRequest.zip](https://github.com/wqweto/VbAsyncSocket/releases/download/HRR-1.0.5/HttpRequest.zip) binary to 1.0.5 release.

**Reply by CouinCouin** [reply]
Hi, I was just trying compilation, I tried MultiUse and GlobalMultiUse, the genretaed dll looks OK (actually testing on a XP virtualbox machine, not at home to test on usual win 10 where I work on VB6 code) : the server displays well the cookie I sent, in the Immediate area. 
I tried also your compiled dll, but I get nothing displayed.
I noticed your DLL has a size of 528 kb, mine 504 kb. 

Edit : In the IDE mine is OK yours not, but in compiled EXE , yours is OK , mine not, do you have an idea ? :)

**Reply by wqweto** [reply]
Could be DLL registration issue. Unregister both DLLs with `regsvr32 /u` and then register the one you want to use in IDE.

I'm releasing digitally signed binaries so the increase of size you are noticing is from the embedded signature.

Edit: This might not be such a good idea because now I remember we had performance problems with certificates on DLLs being checked by Windows which needs access to servers on the network.

**Comment by CouinCouin** [comment]
Hi Vladimir, 
Okay for the dll filesize, I understand :)
Also, I tested to adding directly the 4 files (cAsyncSocket.cls, cHttpRequest.cls, cTlsSocket.cls and mdTlsThunks.bas) and unchecking the "Http Request Replacement" reference to/from the project and it looks running well, on XP and 10. 
Test server seems to receive the cookies I send with "SetCookie" function.
So now, I should not have to register the DLL (with the potential problem you noticed about windows checking certificate) - I assume that with itnegrating classes and module directly, there is not checking anymore(?).
Thanks again for your great help :)

**Reply by wqweto** [reply]
> So now, I should not have to register the DLL

Yes, this way you are linking the component in your final executable.

This is how modern languages work (incl. TwinBasic), by sharing source code components and linking these in final executable.

**Reply by CouinCouin** [reply]
Thanks :) I find also better too like that :)
